### PR TITLE
Properly deserialize dasherized keys

### PR DESCRIPTION
**Pull request by lawitschka**
#### Purpose

The JSON API adapater dasherizes every key, but the deserializer left the keys
unaltered. Thus, the client had to send underscored keys in the request body in
order for Rails to properly match sent values to model attributes.

The changes make overwriting `keyForAttribute()` and `keyForRelationship()` in
Ember.js apps obselete, because Active Model Serializer properly deserializes
the request.

#### Changes

This commit adds automatic key transformation on deserialization. Per default the
deserializer transforms the keys to underscore, but this behaviour can also be
changed by including `key_transform` in the deserializer options.

#### Caveats

I did not include tests for the `KeyTransform#underscore`, because the tests for
key transform only tested the transformation in controller context.

#### Related GitHub issues

https://github.com/rails-api/active_model_serializers/issues/1612

### CODE REVIEW

File: lib/active_model_serializers/adapter/json_api/deserialization.rb [hunk 1]
```diff
@@ -182,23 +182,29 @@ def parse_relationship(assoc_name, assoc_data, options)
           prefix_key = field_key(assoc_name, options).to_s.singularize
           hash =
             if assoc_data.is_a?(Array)
-              { "#{prefix_key}_ids".to_sym => assoc_data.map { |ri| ri['id'] } }
+              { "#{prefix_key}_ids".to_sym => assoc_data.map { |ri| ri[:id] } }
```
Comment (NullVoxPopuli) [inline lib/active_model_serializers/adapter/json_api/deserialization.rb]: why this change?

File: lib/active_model_serializers/key_transform.rb [hunk 2]
```diff
@@ -32,6 +32,16 @@ def dashed(hash)
       hash.deep_transform_keys! { |key| key.to_s.dasherize.to_sym }
     end
 
+    # Transforms keys to underscore.
+    # This is the default case for deserialization in the JsonApi adapter.
+    #
+    # @example:
+    #    "some-key" => "some_key",
+    # @see {https://github.com/rails/rails/blob/master/activesupport/lib/active_support/inflector/methods.rb#L89-L98 ActiveSupport::Inflector.underscore}
```
Comment (NullVoxPopuli) [inline lib/active_model_serializers/key_transform.rb]: thanks for the docs!

File: test/action_controller/json_api/deserialization_test.rb [hunk 3]
```diff
@@ -20,7 +20,10 @@ def test_deserialization
               'id' => 'zorglub',
               'attributes' => {
                 'title' => 'Ember Hamster',
-                'src' => 'http://example.com/images/productivity.png'
+                'src' => 'http://example.com/images/productivity.png',
+                'image-width' => '200',
```
Comment (NullVoxPopuli) [inline test/action_controller/json_api/deserialization_test.rb]: thanks for adding multiple key-format scenarios to the test

File: lib/active_model_serializers/adapter/json_api/deserialization.rb (hunk 1 above)
Comment (lawitschka) [inline lib/active_model_serializers/adapter/json_api/deserialization.rb]: because `KeyTransform#underscore` (and every other method in `KeyTransform`) transforms all keys to symbols, so string key access does not work here

Comment (remear) [inline lib/active_model_serializers/adapter/json_api/deserialization.rb]: Just FYI, this is likely going to change very soon. I've been altering the transform stuff to be more flexible. Stay tuned for that.

**Comment by NullVoxPopuli** [comment]
I'm good with  this, anybody ele,? I'd like. To getiit merged

**Comment by remear** [comment]
Seems good to me.
